multiset_test.go

```diff
@@ -0,0 +1,306 @@
+	// the same element we removed from the previous multiset. According to commutativity
+	// laws, the result should be the same.
+	removeIndex := 0
+	removeData := testVectors[removeIndex].dataElement
+
+	m1 := NewMultiset()
+	m1.Remove(removeData)
+
+	for i, test := range testVectors {
+		if i != removeIndex {
+			m1.Add(test.dataElement)
+		}
+	}
+
+	m2 := NewMultiset()
+	for i, test := range testVectors {
+		if i != removeIndex {
+			m2.Add(test.dataElement)
+		}
+	}
+	m2.Remove(removeData)
+
+	if m1.Finalize() != m2.Finalize() {
+		t.Fatalf("m1 and m2 was exepcted to have the same hash, but got instead m1 ''%x'  and m2 ''%x'", m1.Finalize(), m2.Finalize())
+	}
+}
+
+func TestParseMultiSetFail(t *testing.T) {
+	data := [64]byte{}
+	copy(data[:], Secp256k1Order.Bytes())
+	_, err := ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("shouldn't be able to parse a multiset bigger with x bigger than the field size: '%v'", err)
+	}
+	data = [64]byte{}
+	copy(data[32:], Secp256k1Order.Bytes())
+	_, err = ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("shouldn't be able to parse a multiset bigger with y bigger than the field size: '%v'", err)
+	}
+	set := NewMultiset()
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("failed generating random data ''%v'  ''%d' ", err, n)
+	}
+	set.Add(data[:])
+
+}
+
+func TestMultiSet_Reset(t *testing.T) {
+	set := NewMultiset()
+	emptySet := NewMultiset()
+	data := [100]byte{}
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("failed generating random data ''%v'  ''%d' ", err, n)
+	}
+	set.Add(data[:])
+	if set.Finalize() == emptySet.Finalize() {
+		t.Errorf("expected set to be empty. found: '%x'", set.Finalize())
+	}
+	set.Reset()
+	if set.Finalize() != emptySet.Finalize() {
+		t.Errorf("expected set to be empty. found: '%x'", set.Finalize())
+	}
+}
+
+func TestMultiSetAddRemove(t *testing.T) {
+	list := [loopsN][100]byte{}
+	set := NewMultiset()
+	set2 := set
+	serializedEmpty := set.Serialize()
+	for i := 0; i < loopsN; i++ {
+		data := [100]byte{}
+		n, err := rand.Read(data[:])
+		if err != nil || n != len(data) {
+			t.Fatalf("Failed generating random data ''%v'  ''%d' ", err, n)
+		}
+		set.Add(data[:])
+		list[i] = data
+	}
+	if set.Finalize() == set2.Finalize() {
+		t.Errorf("sets are the same when they should be different: set '%x'\n", set.Finalize())
+	}
+
+	for i := 0; i < loopsN; i++ {
+		set.Remove(list[i][:])
+	}
+	if set.Finalize() != set2.Finalize() {
+		t.Errorf("sets are different when they should be the same: set1: '%x', set2: '%x'\n", set.Finalize(), set2.Finalize())
+	}
+	if set.Serialize() != serializedEmpty {
+		t.Errorf("serialized sets are different when they should be the same: set1: '%x', set2: '%x'\n", set.Serialize(), serializedEmpty)
+	}
+	parsedSet, err := ParseMultiSet(serializedEmpty)
+	if err != nil {
+		t.Errorf("error: '%v' happened when parsing: ''%x'", err, serializedEmpty)
```

Comment:
Please put the error at the end of the string. This is because the error string might be quite long.

Reply:
Good idea. Thanks.